repository_service_tuf/helpers/tuf.py

```diff
@@ -57,6 +85,151 @@ def to_dict(self):
+        tuf_online_key: Key = root_md.signed.keys[online_key_id]
+        online_key_dict = tuf_online_key.to_securesystemslib_key()
+        name = cls._get_key_name(tuf_online_key)
+        online_key = RSTUFKey(online_key_dict, name=name)
+
+        return cls(root_md, online_key)
+
+    def is_keyid_used(self, keyid: str) -> bool:
+        """Check if keyid is used in root keys"""
+        return keyid in self._root_md.signed.roles[Root.type].keyids
+
+    def save_current_root_key(self, key: RSTUFKey):
+        """Update internal information based on 'key' data."""
+        tuf_key: Key = self._root_md.signed.keys[key.key["keyid"]]
+        key.name = self._get_key_name(tuf_key)
+        self.signing_keys[key.name] = key
+
+    def remove_key(self, key_name: str) -> bool:
+        """Try to remove a root key and return status of the operation"""
+        for key in self.keys:
+            if key_name == key["name"]:
+                self._root_md.signed.revoke_key(key["keyid"], Root.type)
+                self.signing_keys.pop(key_name, None)
+                return True
+
+        return False
+
+    def add_key(self, new_key: RSTUFKey) -> None:
+        """Add a new root key."""
+        name = self._get_rstuf_key_name(new_key)
+        new_key.name = name
+        tuf_key = Key.from_securesystemslib_key(new_key.key)
+        tuf_key.unrecognized_fields["name"] = name
+        self._root_md.signed.add_key(tuf_key, Root.type)
+        self.signing_keys[name] = new_key
+
+    def change_online_key(self, new_online_key: RSTUFKey) -> None:
+        """Replace the current online key with a new one."""
+        # Remove the current online key
+        online_key_id = self._root_md.signed.roles["timestamp"].keyids[0]
+        # Top level roles that use the online key
+        online_roles = ["timestamp", "snapshot", "targets"]
+        for role in online_roles:
+            self._root_md.signed.revoke_key(online_key_id, role)
+
+        # Add the new online key
+        name = self._get_rstuf_key_name(new_online_key)
+        new_online_key.name = name
+        online_tuf_key = Key.from_securesystemslib_key(new_online_key.key)
+        online_tuf_key.unrecognized_fields["name"] = name
+        for role in online_roles:
+            self._root_md.signed.add_key(online_tuf_key, role)
+
+        self.online_key = new_online_key
+
+    def has_changed(self) -> bool:
+        """Returns whether the root metadata object has changed"""
+        return self._initial_root_md_obj != self._root_md
+
+    def generate_payload(self) -> Dict[str, Any]:
+        """Save the root metadata into 'file'"""
+        self._root_md.signed.version += 1
+        self._root_md.signatures.clear()
+
+        for key in self.signing_keys.values():
+            self._root_md.sign(SSlibSigner(key.key), append=True)
+
+        console.print("\nVerifying the new payload...")
+        self._root_md.verify_delegate(Root.type, self._root_md)
```

Comment:
> I will need to find a solution to this by storing the original threshold and making sure there is at least a threshold of current keys loaded at the time of new root key generation.

You already have the solution in your RootInfo:
- `_initial_root_md_obj` has the previous threshold and keys
- `_root_md` has the new threshold and keys

Check-out how @kairoaraujo is [doing this in `_trusted_root_update`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L992-L996):

Reply:
minor naming nit: `_initial_root_md_obj` is not ideal. In python-tuf we call the duo "trusted_root" and "new_root". IMO "old" and "new" is also fine. Calling one of them "current" is fine, but a bit ambiguous, because it would fit both. 😄

Reply:
> > I will need to find a solution to this by storing the original threshold and making sure there is at least a threshold of current keys loaded at the time of new root key generation.
> 
> You already have the solution in your RootInfo:
> 
> * `_initial_root_md_obj` has the previous threshold and keys
> * `_root_md` has the new threshold and keys
> 
> Check-out how @kairoaraujo is [doing this in `_trusted_root_update`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L992-L996):

I don't think that's correct. I even think there is a need to fix his implementation as well.
The reason is that the specification clearly says:
`if there is a 1.root.json that has threshold 2 and a 2.root.json that has threshold 3, 2.root.json MUST be signed by at least 2 keys defined in 1.root.json and at least 3 keys defined in 2.root.json` meaning we need to make sure that:
1. there are at least **_current_root_threshold_** a number of keys left in the new root metadata that are also defined in the current (trusted) root metadata.
2. there are at least **_new_root_threshold_** of keys in the new root metadata

What Kairo validates is that `current_root` has a `_current_root_threshold_` of trusted keys and `new_root_threshold` has a `_new_root_threshold_` of keys.
This doesn't make the connection between `_current_root_threshold_`  and `new_root`.

Reply:
I'm pretty sure that Kairo's code does exactly what it should do. It's the same [what ngclient does, when updating root](https://github.com/theupdateframework/python-tuf/blob/e54ee24684e1898cae159484d915ab6a695206be/tuf/ngclient/_internal/trusted_metadata_set.py#L163-L173). You can take a look at the `Metadata.verify_delegate` implementation, if you want to make sure.

Reply:
Ohhh... how could have missed it.
I totally missed that in this line:
[`current_root.verify_delegate(Root.type, new_root)`](https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/33d74e9d4e2333d19b59471fa350ba899aedd855/repository_service_tuf_worker/repository.py#L993)
we are passing the NEW root and thus verifying that the keys in the current root are actually signing the new root and that their amount is enough.
That does make sense.
Thanks for your patience @lukpueh!

Reply:
On that note:

> minor naming nit: _initial_root_md_obj is not ideal. In python-tuf we call the duo "trusted_root" and "new_root". IMO "old" and "new" is also fine. Calling one of them "current" is fine, but a bit ambiguous, because it would fit both. 😄

I think I will use the opportunity to start naming the `current root` to `trusted root` and  `_initial_root_md_obj` will become `trusted_root` internally.

Reply:
> I think I will use the opportunity to start naming the `current root` to `trusted root` and `_initial_root_md_obj` will become `trusted_root` internally.

Wait, you want to call one `trusted root` and the other one `trusted_root`? I don't think that's a good idea.

Reply:
See my latest changes below.